src/PostgREST/AppState.hs

```diff
@@ -91,17 +94,54 @@ initWithPool newPool conf =
     <*> newIORef 0
 
 destroy :: AppState -> IO ()
-destroy = releasePool
+destroy = destroyPool
 
 initPool :: AppConfig -> IO SQL.Pool
 initPool AppConfig{..} =
-  SQL.acquire (configDbPoolSize, configDbPoolTimeout, toUtf8 configDbUri)
-
+  SQL.acquire configDbPoolSize $ toUtf8 configDbUri
+
+-- | Run an action with a database connection.
+--
+-- This is a thin wrapper around 'SQL.use', which deals with a
+-- race on flushing the pool: The pool read via 'readIORef'
+-- might be released via 'flushPool' by the time we call 'SQL.use',
+-- so we catch that 'SQL.PoolIsReleasedUsageError' and retry.
+--
+-- If on the other hand the pool has been (permanently) destroyed,
+-- we re-use 'SQL.PoolIsReleasedUsageError' to convey this to the
+-- caller.
 usePool :: AppState -> SQL.Session a -> IO (Either SQL.UsageError a)
-usePool AppState{..} = SQL.use statePool
-
-releasePool :: AppState -> IO ()
-releasePool AppState{..} = SQL.release statePool
+usePool appState@AppState{..} session = do
+  pool <- readIORef statePool
+  case pool of
+    Nothing -> return $ Left SQL.PoolIsReleasedUsageError
+    Just p -> do
+      res <- SQL.use p session
+      case res of
+        Left SQL.PoolIsReleasedUsageError -> usePool appState session
```

Comment:
This potentially unlimited recursion is a bit scary. I claim that the only situation in which this code path is hit is in the sequence

1. call `usePool`, read `pool1` from the `IORef`
2. `SIGUSR1` hits (or similar), calling `flushPool`
3. `flushPool` creates a new pool `pool2`, switches out the `IORef`, releases `pool1`
4. the `usePool` call from step 1 proceeds to call `SQL.use`, which finds a released pool and thus causes this error.

And now calling `usePool` anew is fine, because we'll necessarily get `pool2` (or later) on the next try.

We could make the code more obviously safe by using `retry` in some form. Or cover it by unit tests.

Reply:
Yeah, true :fearful:. I think at minimum we should log the retry with `AppState.logWithZTime`.